### PR TITLE
feat(api): Add fit and xit with cross-file exclusion

### PR DESCRIPTION
**Pull request by sjbarag**
This PR adds a [jasmine](https://jasmine.github.io/)-like interface to focus on tests and exclude tests from execution via two functions:

* `fit` (short for "focus it"), to run _only_ that test and the other focused tests
* `xit` (short for "exclude it" I guess), to skip execution of a test case

While those sound pretty simple, it requires a fundamental re-structuring of how we execute tests!  We previously executed each test file in isolation, but that doesn't allow a test focused in file A to exclude all the tests in file B.  Bummer!  That's fixed now by making two passes through those test files via `Run` — once to discover tests and once to actually execute the tests that are still enabled.

Also included is better TAP output, based on the expected input to [tapjs's TAP parser](https://github.com/tapjs/tap-parser/blob/d18fbe92f2dd2eaa21cdd480d5811b77bc3cbc7d/index.js).  This allows the `roca` output to be piped into [tap-mocha-reporter](https://github.com/tapjs/tap-mocha-reporter) and for the "spec" reporter (`mocha`'s default) to produce nested suite/sub-suite/test-case hierarchies!

### CODE REVIEW

**Comment by sjbarag** [comment]
Looks like I've still got some docs to write, and — eventually — some tests to write to ensure this code doesn't break.  But this is the direction I want to head!

File: resources/roca_main.brs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    end for
+
+    totalTests = 0
+    for each suite in rootSuites
+        totalTests += suite.__state.totalCases
+    end for
+
+    tap = tap()
+    tap.version()
+    tap.plan(rootSuites.count())
```
Comment (alimnios72) [inline resources/roca_main.brs]: Don't we want to pass the `totalTests` here? or where is being used?

File: resources/tap.brs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+end sub
+
+' Prints a "test case failed" message to the TAP stream
+' @param {integer} index - the zero-based index of the test that failed
+' @param {string} rawTitle - the unsanitized title of the test that failed
+' @param {assocarray} [metadata] - blob of data to be printed in an indented YAML block
+sub __tap_fail(index as integer, rawTitle as string, metadata = invalid as object)
+    print m.getIndent() "not ok " index + 1 " - " m.formatTitle(rawTitle)
+    if metadata <> invalid then
+        ' TODO print a YAML (but probably JSON) document of that data
```
Comment (alimnios72) [inline resources/tap.brs]: Can't the tap reporter write this for us?

File: resources/roca_main.brs (hunk 1 above)
Comment (sjbarag) [reply resources/roca_main.brs]: Surprisingly no!  `tap-parser` seems to want this to be the number of suites.  I think `totalTests` isn't necessary anymore, so I can remove it :)

File: resources/tap.brs (hunk 2 above)
Comment (sjbarag) [reply resources/tap.brs]: Sort of the opposite — `tap-parser` expects YAML and will format it into something else for us.  We might need to do some trickery to get this to work with pretty-printed JSON, but that's a problem for future-us!

File: resources/roca_lib.brs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        if (args.focusedCasesDetected and case.mode = "focus") or not args.focusedCasesDetected then
+            if case.mode <> "skip" then
+                case.exec()
+            end if
+        end if
+        result = case.report(index, tap)
+        if result = "passed" then
+            m.__state.results.passed++
+        else if result = "failed" then
+            m.__state.results.passed++
```
Comment (sjbarag) [inline resources/roca_lib.brs]: 0_o oh no a copy-pasta error!

File: resources/roca_lib.brs [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        __state: m.__state
+    }
+    withM.__func()
+end function
+
+function __case_report(index as integer, tap as object) as string
+    description = buildDescription(m)
+    if m.mode = "skip" or m.__state.success = invalid then
+        tap.skip(index, m.description)
+        return "skiped"
```
Comment (sjbarag) [inline resources/roca_lib.brs]: how embarasing [sic]!

File: resources/roca_lib.brs [hunk 5]
```diff
@@ -0,0 +1,260 @@
+' Creates a suite of test cases with a given description.
+' @param description a string describing the suite of tests contained within.
+' @param func the function to execute as part of this suite.
+' @param
```
Comment (sjbarag) [inline resources/roca_lib.brs]: @sjbarag you accidentally a comment!

File: resources/roca_lib.brs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    if suite.__state.parentSuite = invalid and args.exec = true then
+        suite.exec(args)
+    end if
+
+    return suite
+end function
+
+' Builds a description string for the provided test case that includes all parent suite descriptions.
+' This mimicks the behavior of `mocha` with the `--reporter tap` option.
+function buildDescription(case as object)
```
Comment (sjbarag) [inline resources/roca_lib.brs]: This function is completely unnecessary now!

File: resources/tap.brs (hunk 2 above)
Comment (alimnios72) [reply resources/tap.brs]: Gotcha

File: resources/roca_lib.brs (hunk 5 above)
Comment (alimnios72) [reply resources/roca_lib.brs]: Undocumented params/properties reminds me of some other programming language...

**Pull request merged**